src/kzg10/fastpoly.rs

```diff
@@ -0,0 +1,372 @@
+
+    let mut rev_f = f.clone();
+    let mut rev_g = g.clone();
+    rev_f.reverse();
+    rev_g.reverse();
+
+    let mut q = &rev_f * &inverse_mod_xl::<E>(&rev_g, m + 1).unwrap();
+    q.coeffs.resize(m + 1, E::Fr::zero());
+    rev::<E>(&mut q, m);
+    let r = f - &(g * &q);
+    (q, r)
+}
+
+/// The subproduct tree of a polynomial m over a domain u
+#[derive(Debug, Clone)]
+pub struct SubproductDomain<E: PairingEngine> {
+    /// Domain values
+    pub u: Vec<E::Fr>,
+    /// Subproduct tree over domain u
+    pub t: SubproductTree<E>,
+    /// Derivative of the subproduct polynomial
+    pub prime: P<E>, // Derivative
+}
+
+impl<E: PairingEngine> SubproductDomain<E> {
+    /// Create a subproduct tree domain
+    pub fn new(u: Vec<E::Fr>) -> SubproductDomain<E> {
+        let t = SubproductTree::new(&u);
+        let prime = derivative::<E>(&t.m);
+        SubproductDomain { u, t, prime }
+    }
+    /// evaluate a polynomial over the domain
+    pub fn fast_evaluate(&self, f: &P<E>) -> Vec<E::Fr> {
+        let mut evals = vec![E::Fr::zero(); self.u.len()];
+        self.t.fast_evaluate(f, &self.u, &mut evals);
+        evals
+    }
+    /// interpolate a polynomial over the domain
+    pub fn fast_interpolate(&self, v: &[E::Fr]) -> P<E> {
+        self.t.fast_interpolate(&self.u, v)
+    }
+    /// compute the inverse of the lagrange coefficients fast
+    pub fn fast_inverse_lagrange_coefficients(&self) -> Vec<E::Fr> {
+        self.t.fast_inverse_lagrange_coefficients(&self.u)
+    }
+    /// compute a linear coefficient of lagrange factors times c_i
+    pub fn fast_linear_combine(&self, c: &[E::Fr]) -> P<E> {
+        self.t.fast_linear_combine(&self.u, &c)
+    }
+}
+
+/// A subproduct tree of the subproduct domain
+#[derive(Debug, Clone)]
+pub struct SubproductTree<E: PairingEngine> {
+    /// The left child
+    pub left: Option<Box<SubproductTree<E>>>,
+    /// The right child
+    pub right: Option<Box<SubproductTree<E>>>,
+    /// The polynomial for this subdomain
+    pub m: P<E>,
+}
+
+impl<E: PairingEngine> SubproductTree<E> {
+    /// Compute the subproduct tree of m = (x - u_0)*...*(x-u_{n-1})
+    pub fn new(u: &[E::Fr]) -> SubproductTree<E> {
+        if u.len() == 1 {
+            SubproductTree {
+                left: None,
+                right: None,
+                m: P::<E> {
+                    coeffs: vec![-u[0], E::Fr::one()],
+                },
+            }
+        } else {
+            let n = u.len() / 2;
+            let (u_0, u_1) = u.split_at(n);
+            let left = Box::new(SubproductTree::new(u_0));
+            let right = Box::new(SubproductTree::new(u_1));
+            let m = &left.m * &right.m;
+            SubproductTree {
+                left: Some(left),
+                right: Some(right),
+                m,
+            }
+        }
+    }
+    /// Fast evaluate f over this subproduct tree
+    pub fn fast_evaluate(&self, f: &P<E>, u: &[E::Fr], t: &mut [E::Fr]) {
```

Comment:
Do this (and other methods) need to have the `fast` prefix? Is there a slower alternative?

Reply:
Generally when I prefixed with `fast` I mean that it's quasi-/nearly linear time rather than the straightforward n^2 (or O(nd), in the case of evaluating n arbitrary points on a deg d poly) algorithms. There might be motivation to keep both the fast/slow versions of each operation, as the fast operations require building a `SubproductDomain` first.

Reply:
Right, but these methods are implemented on `SubProductTree`, right? Which means that (by construction), they need to be fast to evaluate.

Reply:
Right, but these methods are implemented on `SubProductTree`, right? Which means that (by construction), they need to be fast to evaluate.

Reply:
oh, I see what you're saying. That make sense, the namespace is already separated by that. right, it should just be evaluate/interpolate/etc.